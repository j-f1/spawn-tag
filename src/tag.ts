import { split } from 'shell-split'
import { SpawnOptions } from 'child_process'
import crossSpawn = require('cross-spawn')
import pEvent = require('p-event')

export type TemplateTag<T> = (
  strings: TemplateStringsArray,
  ...interpolations: unknown[]
) => T

export interface Options extends SpawnOptions {
  capture?: {
    stdout?: BufferEncoding | boolean
    stderr?: BufferEncoding | boolean
  }
}

export default function tag<T extends Options>(options: T) {
  const opts: T & Required<Pick<Options, 'capture'>> = Object.assign(
    { capture: { stdout: false, stderr: false } },
    options,
  )
  type Result = {
    stdout: string | Buffer | null
    stderr: string | Buffer | null
  }

  const tag: TemplateTag<Promise<Result>> = (strings, ...interpolations) => {
    let sep = '%expr%'
    while (strings.find(str => str.includes(sep))) {
      sep += '%'
    }
    const args = [...split(strings.join(sep))]
    let i = 0
    for (let idx = 0; idx < args.length; idx++) {
      while (args[idx].includes(sep)) {
        args[idx] = args[idx].replace(sep, () => String(interpolations[i++]))
      }
    }

    const [cmd, ...argv] = args

    const childProcess = crossSpawn(cmd, argv, opts)

    const result: Result = {
      stdout:
        opts.capture.stdout != null
          ? typeof opts.capture.stdout === 'string'
            ? ''
            : Buffer.alloc(0)
          : null,
      stderr:
        opts.capture.stderr != null
          ? typeof opts.capture.stderr === 'string'
            ? ''
            : Buffer.alloc(0)
          : null,
    }

    if (opts.capture.stdout != null) {
      childProcess.stdout.on('data', (data: Buffer) => {
        if (typeof opts.capture.stdout === 'string') {
          result.stdout += data.toString(opts.capture.stdout)
        } else if (result.stdout instanceof Buffer) {
          result.stdout = Buffer.concat([result.stdout, data])
        }
      })
    }
    if (opts.capture.stderr != null) {
      childProcess.stderr.on('data', (data: Buffer) => {
        if (typeof opts.capture.stderr === 'string') {
          result.stderr += data.toString(opts.capture.stderr)
        } else if (result.stderr instanceof Buffer) {
          result.stderr = Buffer.concat([result.stderr, data])
        }
      })
    }

    const onExit = pEvent<[number, string]>(childProcess, 'exit', {
      multiArgs: true,
    })
    return Object.assign(
      onExit.then(arg => {
        const [code, signal] = arg
        if (code || signal) {
          throw Object.assign(
            new Error(
              `Process exited with ${
                code ? `code ${code}` : `signal ${signal}`
              }.`,
            ),
            { code, signal },
          )
        }
        return result
      }),
      { childProcess },
    )
  }
  return tag
}
