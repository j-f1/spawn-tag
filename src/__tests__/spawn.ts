import spawn, { Options } from '..'

import { TemplateTag } from '../tag'

import { join, relative } from 'path'

///<reference types="jest" />

const fixturePath = (...args: string[]) =>
  relative(process.cwd(), join(__dirname, 'fixtures', ...args))

const t = (name: string, opts?: Options): TemplateTag<void> => (
  strings,
  ...args
) => {
  it(name, async () => {
    const callee = opts ? spawn(opts) : spawn
    expect(await callee(strings, ...args)).toMatchSnapshot()
  })
}

describe('The `spawn` tag', () => {
  t('runs `echo` without errors')`echo hello`

  t('handles interpolations with spaces properly')`echo ${'Hello, world'}`

  t('converts non-string interpolations to strings')`echo ${33}`

  t('escapes semicolons')`echo ${'Hello!; echo Goodbye!'}`

  t('runs `ls` properly')`ls ${fixturePath('dir')}`

  t('does not quote interpolated options')`ls ${'-g'} ${fixturePath('dir')}`

  it('passes space-delimited text as one option', async () => {
    const promise = spawn`cat ${'- ' + fixturePath('file')}`
    promise.childProcess.stdin.end('hello')
    expect(
      await promise.catch(arg => JSON.parse(JSON.stringify(arg))),
    ).toMatchSnapshot()
  })

  t(
    'handles interpolations adjacent to literal text properly',
  )`echo abc${'def'}`

  t('handles `spawn` options correctly', {
    env: { ...process.env, FOO: 'foo' },
  })`node -e 'console.log(process.env.FOO)'`
})

describe('The `spawn.silently` tag', () => {
  it('runs `echo` without errors', async () => {
    expect(await spawn.silently`echo hello`).toMatchSnapshot()
  })
})
