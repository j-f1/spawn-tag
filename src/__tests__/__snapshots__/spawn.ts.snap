// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The \`spawn.silently\` tag runs \`echo\` without errors 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": null,
  "stdout": null,
}
`;

exports[`The \`spawn\` tag converts non-string interpolations to strings 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "33
",
}
`;

exports[`The \`spawn\` tag does not quote interpolated options 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "total 0
-rw-r--r--  1 staff  0 Sep 30 08:04 bar
-rw-r--r--  1 staff  0 Sep 30 08:04 foo
",
}
`;

exports[`The \`spawn\` tag escapes semicolons 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "Hello!; echo Goodbye!
",
}
`;

exports[`The \`spawn\` tag handles \`spawn\` options correctly 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "foo
",
}
`;

exports[`The \`spawn\` tag handles interpolations adjacent to literal text properly 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "abcdef
",
}
`;

exports[`The \`spawn\` tag handles interpolations with spaces properly 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "Hello, world
",
}
`;

exports[`The \`spawn\` tag passes space-delimited text as one option 1`] = `
Object {
  "code": 1,
  "signal": null,
  "stderr": "cat: illegal option --  
usage: cat [-benstuv] [file ...]
",
  "stdout": "",
}
`;

exports[`The \`spawn\` tag runs \`echo\` without errors 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "hello
",
}
`;

exports[`The \`spawn\` tag runs \`ls\` properly 1`] = `
Object {
  "code": 0,
  "signal": null,
  "stderr": "",
  "stdout": "bar
foo
",
}
`;
